library/cc/bridge_utility.cc

```diff
@@ -0,0 +1,115 @@
+#include "bridge_utility.h"
+
+#include <sstream>
+
+namespace Envoy {
+namespace Platform {
+
+// TODO(crockeo): we always copy memory across boundaries; consider allowing for moves and/or
+// shared ownership w/ reference counting via envoy-mobile's release callbacks
+
+envoy_data buffer_as_envoy_data(const std::vector<uint8_t>& data) {
+  size_t byte_len = sizeof(uint8_t) * data.size();
+
+  uint8_t* bytes = static_cast<uint8_t*>(safe_malloc(byte_len));
+  memcpy(bytes, &data[0], byte_len);
+  envoy_data raw_data{
+      .length = byte_len,
+      .bytes = bytes,
+      .release = envoy_noop_release,
+      .context = nullptr,
+  };
+
+  return raw_data;
+}
+
+envoy_data string_as_envoy_data(const std::string& data) {
+  size_t byte_len = sizeof(uint8_t) * data.size();
+
+  uint8_t* bytes = static_cast<uint8_t*>(safe_malloc(byte_len));
+  memcpy(bytes, &data[0], byte_len);
+  envoy_data raw_data{
+      .length = byte_len,
+      .bytes = bytes,
+      .release = envoy_noop_release,
+      .context = nullptr,
+  };
+
+  return raw_data;
+}
+
+envoy_headers raw_headers_as_envoy_headers(const RawHeaders& headers) {
+  envoy_header* headers_list =
+      static_cast<envoy_header*>(safe_malloc(sizeof(envoy_header) * headers.size()));
```

Comment:
I think this is maybe the wrong size - since the number of entries in the map won't account for header names that have multiple values.